elastic/logstash
# Feature: health report api

## Release notes

Adds a new health report API (`GET /_health_report`), with indicators that are capable of detecting degraded status of configured pipelines and providing actionable insight.

## What does this PR do?

 - Adds `GET /_health_report` endpoint with indicators capable of diagnosing degraded status of configured pipelines, in accordance with the [Health Report RFC](https://github.com/elastic/logstash/issues/16056).

The feature-branch that we are requesting to merge is a combination of several previously-reviewed pull-requests:
 - #16398: health report endpoint
 - #16467: health report integration tests
 - #16518: health report api docs, diagnosis links
 - #16141: bootstrap health observer 

## Why is it important/What is the impact to the user?

 - Provides a channel for actionable insight into the runtime state of pipelines
 - This adds the base health report endpoint in a way that is extensible and will allow us to layer on additional actionable insights in future releases

## Checklist




## How to test this PR locally

 - build logstash locally (`./gradlew clean assemble installDefaultGems`)
 - download and extract the demo [demo.tgz](https://github.com/user-attachments/files/17298890/demo.tgz)
 - set `LOGSTASH_HOME` and run the demo (e.g., `LOGSTASH_HOME="${PWD}/.." ./demo.sh`)
 - follow the on-screen prompts (you'll need to hit `<ENTER>` when you see a key icon, and to `<CTRL>+C` _once_ to get out of each `watch` screen)


## Use cases

 - avoid including a pipeline in upstream LB until it is running
 - know when a pipeline has crashed
 - know when a pipeline has finished

## Screenshots

![Screenshot 2024-10-08 at 1 27 57 PM](https://github.com/user-attachments/assets/7f4a950f-16e7-4e33-bbb8-c54d74e65d64)
![Screenshot 2024-10-08 at 1 29 16 PM](https://github.com/user-attachments/assets/335cba90-9ec5-42b0-87c8-0f892ff60b83)
![Screenshot 2024-10-08 at 1 29 45 PM](https://github.com/user-attachments/assets/9e09ff24-57ee-43d4-b0cd-79afb29d14eb)
![Screenshot 2024-10-08 at 1 30 20 PM](https://github.com/user-attachments/assets/b1886b2b-2352-4844-b74a-dd00f726a21b)
![Screenshot 2024-10-08 at 1 30 42 PM](https://github.com/user-attachments/assets/1c29e73d-88cf-43d1-8efe-f1214b717a5d)
